### PR TITLE
Friendship model #v1

### PR DESCRIPTION
**Pull request by blackpintz**
## The user can:

- Send a friendship invitation using a button on the users' list page and single-user page.

- Receive a friendship invitation and view it on the users' list page and single-user page.

- Accept or Ignore a friendship invitation from the users' list page and single-user page.

- View their posts and friends' posts on their timeline.

### CODE REVIEW

**Comment by blackpintz** [comment]
It shows checks failed but no linter errors are found by stickler-ci

File: app/views/users/show.html.erb [hunk 1]
```diff
@@ -1,9 +1,49 @@
...
+    <p>Your request from <%= @user.name %> is pending.</p>
+       <span>
+        <%= link_to 'Accept', friendship_path(@user, user_id: @user.id, friend_id: current_user.id), method: :put, class: 'frnd-btn' %>
+      </span>
+      <span>
+          <%= link_to 'Ignore', friendship_path(@user, user_id: @user.id, friend_id: current_user.id), method: :delete, class: 'ignore-btn' %>
+      </span>
+  </div>
+  <% end %>
+  <% unless request_status(current_user,@user) || @user == current_user %>
```
Comment (wathigo) [inline app/views/users/show.html.erb:14]: **REQUIRED**
* There should absolutely no logic in the views. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright).
[The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

File: app/views/users/show.html.erb [hunk 2]
```diff
@@ -1,9 +1,49 @@
 <section class="user-section">
   <h2><%= "Name: #{@user.name}." %></h2>
+  <% if show_requests_on_userlist(@user, current_user) %>
```
Comment (wathigo) [inline app/views/users/show.html.erb:3]: **REQUIRED**
* There should absolutely no logic in the views. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright).
[The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

File: app/views/users/show.html.erb [hunk 3]
```diff
@@ -1,9 +1,49 @@
...
+    </span>
+  <% end %>
   <article class="timeline">
     <h3>Recent posts:</h3>
     <ul class="posts">
       <%= render @posts %>
     </ul>
   </article>
+  <div>
+    <% if @user == current_user %>
```
Comment (wathigo) [inline app/views/users/show.html.erb:26]: **REQUIRED**
* There should absolutely no logic in the views. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright).
[The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

File: app/views/users/show.html.erb [hunk 4]
```diff
@@ -1,9 +1,49 @@
...
+          <%= link_to 'Ignore', friendship_path(@user, user_id: friend.user.id, friend_id: friend.friend.id), method: :delete, class: 'ignore-btn' %>
+      </span>
+    <% end %>
+    <% end %>
+  </div>
+  <div>
+    <h4><%= @user.name %>'s friends:</h4>
+    <ul>
+    <% @users.each do |user| %>
+    <% if check_friend_status(@user, user) %>
```
Comment (wathigo) [inline app/views/users/show.html.erb:43]: **REQUIRED**
* There should absolutely no logic in the views. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright).
[The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

File: app/helpers/application_helper.rb [hunk 5]
```diff
@@ -15,4 +15,31 @@ def like_or_dislike_btn(post)
       link_to('Like!', post_likes_path(post_id: post.id), method: :post)
     end
   end
+
+  def show_requests_on_userlist(user, friend)
+    friend = Friendship.where(user_id: user, friend_id: friend, status: false)
+    !!friend.first
+  end
```
Comment (wathigo) [inline app/helpers/application_helper.rb:22]: **REQUIRED**
* There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...).
Everything connected with data processing should be encapsulated in models and used in controllers.
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

File: app/helpers/application_helper.rb [hunk 6]
```diff
@@ -15,4 +15,31 @@ def like_or_dislike_btn(post)
...
+
+  def pending_requests(current, user)
+    requests = Friendship.where(user_id: current.id, friend_id: user.id)
+    !!requests.first
+  end
```
Comment (wathigo) [inline app/helpers/application_helper.rb:27]: **REQUIRED**
* There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...).
Everything connected with data processing should be encapsulated in models and used in controllers.
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -15,4 +15,31 @@ def like_or_dislike_btn(post)
...
+
+  def request_status(current, user)
+    pending_requests(current, user) || pending_requests(user, current)
+  end
+
+  def accepted_requests(current, user)
+    friend = Friendship.where(user_id: current.id, friend_id: user.id, status: true)
+    !!friend.first
+  end
```
Comment (wathigo) [inline app/helpers/application_helper.rb:36]: **REQUIRED**
* There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...).
Everything connected with data processing should be encapsulated in models and used in controllers.
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

**Comment by uryelah** [comment]
Hi **Rose Wanjohi**, I'm Sarah your reviewer for this PR 👩‍🚀 

Great job with the fixes! You are almost there, there's just some code organization issues remaining and then you'll be all good with this milestone! Keep your code game strong! 💪 

# Required Changes 🏗️ 🐛 💡

- [ ] There's still some view logic remaining in the views, check the conditional at the posts index for example.

- [ ][OPTIONAL]You could work on some re-usable partials for the post form and the user show and index views to avoid repeating lines.

- [ ] There should be no ActiveRecord queries(where, order, join, etc) and logic directly in the controllers. 
For the timeline_posts for example, all those .where queries should be coming from a model method.


-----

 **Kindly fix the pointed out issues before requesting the next code review**

------------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

**Pull request merged**